src/Engine/ProtoCore/FFI/CLRObjectMarshaler.cs

```diff
@@ -1105,20 +1108,27 @@ private StackValue CreateDSObject(object obj, ProtoCore.Runtime.Context context,
         {
             //We are here, because we want to create DS object of user defined type.
             var runtimeCore = dsi.runtime.RuntimeCore;
-            var classTable = runtimeCore.DSExecutable.classTable;
             Type objType = GetPublicType(obj.GetType());
-            int type = classTable.IndexOf(GetTypeName(objType));
-            //Recursively get the base class type if available.
-            while (type == -1 && objType != null)
+
+            if (mCachedObjType != objType)
             {
-                objType = objType.BaseType;
-                if (null != objType)
-                    type = classTable.IndexOf(GetTypeName(objType));
+                mCachedObjType = objType;
+
+                var classTable = runtimeCore.DSExecutable.classTable;
+                mCachedType = classTable.IndexOf(GetTypeName(objType));
```

Comment:
is this implemented with a map - or does look through the entire class table?

Reply:
Nope.... is is trying to get an exact match first with Linq.  Next it tries to get a match by splitting the the string also with Linq.